src/meta/src/barrier/creating_job/barrier_control.rs

```diff
@@ -228,46 +175,45 @@ impl CreatingStreamingJobBarrierControl {
             .set(self.inflight_barrier_queue.len() as _);
     }
 
-    #[expect(clippy::type_complexity)]
-    /// Return (`upstream_epochs_to_notify`, Some((epoch, resps, `is_first_commit`)))
+    /// Return Some((epoch, resps, `is_first_commit`))
     ///
-    /// `upstream_epochs_to_notify` is the upstream epochs of non-checkpoint barriers to be notified about barrier completing.
-    /// These non-checkpoint barriers does not need to call `commit_epoch` and therefore can be completed as long as collected.
+    /// Only epoch within the `epoch_end_bound` can be started.
     pub(super) fn start_completing(
         &mut self,
-    ) -> (Vec<u64>, Option<(u64, Vec<BarrierCompleteResponse>, bool)>) {
-        if self.completing_barrier.is_some() {
-            return (vec![], None);
-        }
-        let mut upstream_epochs_to_notify = vec![];
-        while let Some(mut epoch_state) = self.pending_barriers_to_complete.pop_back() {
+        epoch_end_bound: Bound<u64>,
```

Comment:
What is this `epoch_end_bound` referring to, could you add some docs for it?

Reply:
It is clear from the function header, that it has this property:
```
Only epoch within the `epoch_end_bound` can be started.
```

But it's not clear why we need it, what's its function.

Reply:
Added a comment. 
```
/// Usually `epoch_end_bound` is the upstream committed epoch. This is to ensure that
/// the creating job won't have higher committed epoch than the upstream.
```